Adds FX3 colorization support

GyroJoe opened the pull request:
Adds support for colorizing FX3 frames when using the memory grabber.

To achieve this, a new memory grabber has been added that can determine the name of the current game. (It appears to be the name for the UI skin the menu uses). These names are the same as the ones used for backglass images (e.g. `WMS_Monster_Bash`).

Once the game name is determined, the palette is loaded, and switched to dynamically. In this way, new colorizations can be loaded without tearing down the entire render graph.

The same `altcolor` folder location strategy is used as for VPinMame. It also now searches the COM location, using the same trick the installer does.

For now, I only added support for PAL/VNI, but Serum should be feasible - need a colorization for a supported game to test with. I know there has been some churn in the colorization space lately - I tried to avoid changing the logic as much as possible and only moved things to a new `ColorizationLoader` so the code could be shared. Hopefully this is not disruptive to any on-going refactor work.

GyroJoe commented:
Quick screenshot of it in action:
![image](https://user-images.githubusercontent.com/2504702/232267588-ca0362d9-59b7-4d2b-9fac-3cecc5357911.png)

GyroJoe commented on LibDmd/Converter/SwitchingConverter.cs:
```diff
@@ -0,0 +1,128 @@
+	/// Converter that can swap in another child converter dynamically, without requiring upstream subscribers to re-subscribe.
+	/// Falls back to orange DMD colored frames when no child converter is available.
+	/// </summary>
+	public class SwitchingConverter : AbstractSource, IConverter, IColoredGray2Source, IColoredGray4Source, IColoredGray6Source
```
I think all of this is correct, but this is probably the code I'm the least sure of - could use another set of eyes.

freezy replied:
Apart from the colors of the pass-through, it looks okay to me as well.

zesinger commented:
That's really amazing! Nice addition, thanks! I'd be happy to help for Serum,

freezy commented:
Dude, you're killing it! Thanks!

I haven't started the coloring refactoring yet, so I'll review and merge this one first. Cheers!

freezy commented:
Had a quick test. Nice!

![image](https://user-images.githubusercontent.com/70426/232326037-727d0958-ebe2-49b5-b44a-95580cb23767.png)

I'll finish the review later today and push some minor changes. Would be awesome to have this integrated with the other sources as well (FP, PBA).

freezy commented on LibDmd/Converter/SwitchingConverter.cs:
```diff
@@ -0,0 +1,112 @@
+			}
+			else
+			{
+				if (frame.BitLength == 4) {
+					_coloredGray4PassthroughFrames.OnNext(new ColoredFrame(frame.width, frame.height, frame.Data, Color.FromRgb(0xff, 0x66, 0x00)));
+				} else {
+					_coloredGray2PassthroughFrames.OnNext(new ColoredFrame(frame.width, frame.height, frame.Data, Color.FromRgb(0xff, 0x66, 0x00)));
+				}
```
I've added passthrough for 4-bit frames as well (PinMAME has support for both 2-bit and 4-bit games).

However, the color's still wrong here. Before, the FX3 grabber was a `IColoredGray2Source`, since the hue of the DMD could change from game to game. Now, it's an `IGray2Source`, so we'd need to pass `PinballFX3MemoryGrabber._dmdColor` into the `ColoredFrame` somehow.

GyroJoe replied:
Yea, I had a looked at this a bit - it seemed like the dmdColor part of the memory inspection wasn't working at all.

I poked around a bit but couldn't find the new memory address.

freezy replied:
Huh, okay. I'll have a look tomorrow, but finding memory addresses isn't really my thing neither.

GyroJoe replied:
I got it figured out actually - it was only off by one word.

Added an Observable for the color, so it can dynamically swap.

freezy commented on LibDmd/Input/PinballFX/PinballFX3MemoryGrabber.cs:
```diff
@@ -11,11 +11,11 @@ namespace LibDmd.Input.PinballFX
 	/// launched and stop sending when it exits.
 	/// </remarks>
-	public class PinballFX3MemoryGrabber : MemoryGrabber<ColoredFrame>, IColoredGray2Source
+	public class PinballFX3MemoryGrabber : PinabllFX3MemoryGrabberBase<DMDFrame>, IGray2Source
```
There's a typo here: ->`PinballFX3MemoryGrabberBase`

freezy reviewed (commented):
This is great. A few more things before we merge this:

- [ ] Fix the color hue when coloring is disabled, and enabled but no colorization available
- [ ] Add the same feature to Future Pinball and Pinball Arcade
- [ ] Document the feature in README.md
- [x] Test Serum and PinMAME

I'll have a go at the second and last point.

freezy commented:
Serum now working.

![image](https://user-images.githubusercontent.com/70426/232339613-f78918b1-aea6-4cec-8326-4e9ebd2fbfa9.png)

zesinger commented:
Amazing! Thanks

GyroJoe commented on README.md:
```diff
@@ -140,6 +140,20 @@ The DMD from Pinball FX3 is pulled directly from the memory.
 It doesn't matter whether Pinball FX3 is started before or after `dmdext`, and
 it works with or without cabinet mode.
 
+Like VPM, the FX3 memory grabber also supports colorization on a monitor or RGB display.
```
I'm wondering if we should move all the info on colorization to their own section, rather than having it under each game, since there is a fair amount of overlap.

freezy replied:
Yes, that's a good idea. Also, add a table of contents :)

freezy commented:
Pinball Arcade is somewhat working, but I had to change the frame format, so 4-bit games don't work right now. Will see if I find a fix and push tomorrow.

![image](https://user-images.githubusercontent.com/70426/232347371-fbe7ae68-695f-4c1f-a3db-f7bdc1a96587.png)

freezy commented:
Oh, and I just created an `IGameNameSource` interface and made PBA implement it. @GyroJoe is there any reason you've created a completely new grabber for the name, instead of adding this to the existing one?

GyroJoe commented:
I mainly didn't love the idea of that one grabber doing so many different things. It's also running at a different framerate (1 FPS, lol), since the game name is infrequently changing.

GyroJoe commented:
Should fix #344, and at least the color part of #258

freezy commented:
> I mainly didn't love the idea of that one grabber doing so many different things. It's also running at a different framerate (1 FPS, lol), since the game name is infrequently changing.

Okay. I might move it into the original grabber, because a) it allows to abstract the name retrieval across sources and b) have less duplicated code. I'll measure the performance penalty, but my guess is that's it's negligible.

freezy commented:
I've pushed the TPA update. @GyroJoe let me know what you think. I didn't touch FX3 yet, but basically you first set up the graphs, and at the end the switching converter, if the source supports it.

GyroJoe commented:
> I've pushed the TPA update. @GyroJoe let me know what you think. I didn't touch FX3 yet, but basically you first set up the graphs, and at the end the switching converter, if the source supports it.

Clean, and unifies the setup across sources. I implemented this for FX3 and added a new type `IDmdColorSource`.

One wrinkle with the unified grabber for FX3 is that the game name and the DMD aren't both available at the same point in time. I added a one-time second attempt to get the name if it fails the first time through. This was one advantage of the previous split system I guess, so I never noticed.

freezy commented:
Awesome, thanks a lot! Looks much cleaner indeed.

Also thanks for updating `.editorconfig`, I might tweak it a little more.

freezy commented:
So the remaining tasks are Future Pinball, although it's currently only emitting 4-bit frames, which I'd have to convert, and I don't know how accurate the animations and texts are so they'd be recognized by the coloring plugins (FP tables aren't ROM based, so authors re-create the DMD output manually). I'll also spend some time on reorganizing the documentation a little.

@GyroJoe you don't happen to be interested in working on a new pinball simulator based on Unity? ;)

GyroJoe commented:
> Also thanks for updating `.editorconfig`, I might tweak it a little more.

A word of warning, the editor in VS for this file is pretty broken. Just change it with a normal text editor.
https://github.com/dotnet/roslyn/issues/54556

freezy commented:
For me it's ready to merge now. @GyroJoe if you wanna have a last look, lemme know, otherwise I'll go ahead.

GyroJoe commented:
Wow, you went the extra mile with the docs including all the game names. 

🚢 it

lucky01 commented:
@GyroJoe Great new feature ! Thanks for that !
@frezzy I would appreciate instead of making a release with this new feature based on deprecated vni/pal support, to make the next release together with the refactored plugin interface and enable pac support that way. Otherwise it encourages the community to continue to illegally share vni/pal colorization against the will of the authors. We should not repeat history here.

freezy commented:
@lucky01 That's an absurd suggestion. This feature was already announced on several Discords, how long do you think it'll take to spread on the forums anyway? It's an open source project, remember.
